File: GC.WebSpace/ClientApp/domain/users/userBlank.ts
import { User } from "./user";
import { UserPermissionBlank } from "./userPermissionBlank";

export class UserBlank {
    constructor(
        public id: string | null,
        public login: string,
        public password: string,
        public rePassword: string,
        public passwordWasEdited: boolean,
        public permissions: UserPermissionBlank[]
    ) { }

    public static getDefault(): UserBlank {
        return new UserBlank(null, '', '', '', false, []);
    }

    public static create(user: User): UserBlank {
        return new UserBlank(user.id, user.login, '', '', false, []);
    }

    public get isValid() : boolean{
        if (this.id == null || this.password != ""){
            return  this.login != null && this.login != "" &&
                    this.password != null && this.password != "" &&
                    this.rePassword != null && this.rePassword != "" &&
                    this.permissions != null && this.permissions.length > 0
        } else{
            return this.login != null && this.login != "" &&
            this.permissions != null && this.permissions.length > 0

        }
    }

}

File: GC.Domain/Users/UserBlank.cs
using GC.Tools.Extensions;
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;

namespace GC.Domain.Users
{
    public class UserBlank : IValidatableObject
    {
        public Guid? Id { get; set; }
        public String? Login { get; set; }
        public String? Password { get; set; }
        public String? RePassword { get; set; }
        public Boolean? PasswordWasEdited { get; set; }
        public UserPermissionBlank[] Permissions { get; set; }

        public IEnumerable<ValidationResult> Validate(ValidationContext validationContext)
        {
            List<ValidationResult> errors = new List<ValidationResult>();

            if (String.IsNullOrWhiteSpace(Login)) errors.AddError("Логин пуст, либо содержит некорректное значение");

            if(Id is null)
            {
                if (String.IsNullOrWhiteSpace(Password)) errors.AddError("Пароль пуст, либо содержит некорретное значение");
                if (String.IsNullOrWhiteSpace(RePassword)) errors.AddError("Повторение пароля пусто, либо содержит некорретное значение");
                if (Password != RePassword) errors.AddError("Пароли не совпадают");
            }
            else
            {
                if (!String.IsNullOrEmpty(Password) || !String.IsNullOrEmpty(RePassword))
                {
                    if (Password != RePassword) errors.AddError("Пароли не совпадают");
                }
            }

            if (Permissions is null || Permissions.Length == 0) errors.AddError("Не указан список ролей пользователя");

            return errors;
        }
    }
}
